sklearn/utils/estimator_checks.py

```diff
@@ -333,42 +334,100 @@ def _construct_instance(Estimator):
     return estimator
 
 
+def _get_estimator_name(estimator):
+    """Get name of estimator instance or class"""
+    if isinstance(estimator, type):
+        # got a class
+        return estimator.__name__
+    # got an instance
+    return type(estimator).__name__
+
+
+def _get_xfail_checks(estimator):
+    """Get checks marked with xfail_checks tag from estimator"""
+    if isinstance(estimator, type):
+        # try to construct estimator instance, if it is unable to then
+        # the xfail_checks tag is ignored
+        try:
+            estimator = _construct_instance(estimator)
+        except Exception:
+            return {}
+
+    return estimator._get_tags()['_xfail_checks'] or {}
+
+
+def _make_check_warn_on_fail(estimator, check, xfail_checks_tag):
+    """Wrap the check so that a warning is raised:
+
+    - when the check is in the `xfail_checks` tag and the check properly failed
+      as expected
+    - when the check is in the `xfail_checks` tag and the check didnt fail but
+      should have
+
+    Checks that aren't in the xfail_checks tag aren't wrapped and are returned
+    as-is. This wrapper basically simulates what pytest would do with the
+    @xfail decorator, but this one can be used with check_estimator() which
+    doesn't rely on pytest.
+
+    """
+    check_name = _set_check_estimator_ids(check)
+    if check_name not in xfail_checks_tag:
+        return estimator, check
+
+    reason = xfail_checks_tag[check_name]
+    est_name = _get_estimator_name(estimator)
+    @wraps(check)
+    def wrapped(*args, **kwargs):
+        try:
+            check(*args, **kwargs)
+        except AssertionError:
+            warnings.warn(reason, SkipTestWarning)
```

Comment:
I find it a bit strange to raise a SkipTestWarning when the test wasn't actually skipped: the check was run, but we got a failure as expected.

Reply:
I do not think we have a good warning for "expected to fail".

Reply:
Maybe a fairly ambiguous "UserWarning".